src/index.js

```diff
@@ -1,44 +1,37 @@
-"use strict";
+'use strict';
 
-var isInteger = require("is-integer");
-var matrix = require('ml-matrix');
-
-/**
- * Savitzky-Golay filter
- * @param {Array <number>} data
- * @param {number} h
- * @constructor
- */
-function SavitzkyGolay (data, h) {
-    this.data = data;
-    this.h = h;
-}
+var Matrix = require('ml-matrix');
+var padArray = require('ml-pad-array');
+var extend = require('extend');
 
 var defaultOptions = {
     windowSize: 5,
     derivative: 1,
-    polynomial: 2
+    polynomial: 2,
+    padValue: 0
 };
 
 /**
- * Applies the filter
+ * Savitzky-Golay filter
+ * @param {Array <number>} data
+ * @param {number} h
  * @param {Object} options
  * @returns {Array}
  */
-SavitzkyGolay.prototype.calc = function (options) {
-    options = options || {};
-    for (var o in defaultOptions)
-        if (!(options.hasOwnProperty(o)))
-            options[o] = defaultOptions[o];
-    if ((options.windowSize % 2 === 0) || (options.windowSize < 5) || !(isInteger(options.windowSize)))
+function SavitzkyGolay (data, h, options) {
+    options = extend({}, defaultOptions, options);
+    if ((options.windowSize % 2 === 0) || (options.windowSize < 5) || !(Number.isInteger(options.windowSize)))
         throw new RangeError('Invalid window size (should be odd and at least 5 integer number)');
-    if ((options.derivative < 0) || !(isInteger(options.derivative)))
+    if ((options.derivative < 0) || !(Number.isInteger(options.derivative)))
         throw new RangeError('Derivative should be a positive integer');
-    if ((options.polynomial < 1) || !(isInteger(options.polynomial)))
+    if ((options.polynomial < 1) || !(Number.isInteger(options.polynomial)))
         throw new RangeError('Polynomial should be a positive integer');
 
     var C, norm;
-    var ans =  new Array(this.data.length);
+    var step = Math.floor(options.windowSize / 2);
+    data = padArray(data, {size: step, value: options.padValue});
```

Comment:
To me, it seems to be the expected behavior.

If we take the example from #10: `[ 1, 2, 3, 4, 5, 4, 3, 2, 1 ]`
with default prepadding, we have `[ 0.8, 1, 1, 0.6, 0, -0.6, -1, -1, -0.8 ]`
with postpadding, we have `[ 0, 0, 1, 0.6, 0, -0.6, -1, 0, 0 ]`

Comparing with matlab (derivative: 0, polynomial: 1):

prepadding: `[ 1.2, 2, 3, 3.6, 3.8, 3.6, 3, 2, 1.2 ]`
postpadding: `[ 0, 0, 3, 3.6, 3.8, 3.6, 3, 0, 0 ]`
matlab: `[ 1, 2, 3, 3.6, 3.8, 3.6, 3, 2, 1 ]`

I think that matlab just doesn't care about the borders and copies them in the result. We could add such an option later, but anyway I don't see a reason why we would like to lose the border values.